service_linux.go

```diff
@@ -299,13 +299,22 @@ func filterPortConfigs(ingressPorts []*PortConfig, isDelete bool) []*PortConfig
 
 func programIngress(gwIP net.IP, ingressPorts []*PortConfig, isDelete bool) error {
 	addDelOpt := "-I"
+	rollbackAddDelOpt := "-D"
 	if isDelete {
 		addDelOpt = "-D"
+		rollbackAddDelOpt = "-I"
 	}
 
 	ingressMu.Lock()
 	defer ingressMu.Unlock()
 
+	handleIngressPorts := false
+	defer func() {
+		if !handleIngressPorts {
+			filterPortConfigs(ingressPorts, !isDelete)
+		}
+	}()
+
```

Comment:
My reading of this is that if we get past the setup phase (the next block) then we remove the ingress ports that we were trying to add.   In the first place, this seems wrong since it does this whether rest of the function succeeds or fails.  But more importantly ... I think that the `filterPortConfigs()` aspect of your original patch was the more correct way to handle this part of the rollback.  i.e. if `programIngress()` returns an error during an add operation, then run `filterPortConfigs(lb.service.ingressPorts, false)` in the error block.  The only thing wrong with that part of the original was that it was incomplete.  The rest of your patch sees to undoing the iptables programming.